### ISSUE TITLE
Better line info for "missing comma or ) in argument list" error

### ISSUE THREAD

**Issue by KristofferC** [JuliaLang/julia#16163]
Putting this in a file and `include` it:

``` jl
function filler()
end

function g(a,b,c)
    f(a,b)
    f(a,b)
    f(a,(b,))
    f(a,(b,))
    f(a,(b,))
    f(a,(b,) # :(
    f(a,(b,))
    f(a,(b,))
    f(a,(b,))
end
```

gives the error message:

``` jl
julia> include("test.jl")
ERROR: LoadError: syntax: missing comma or ) in argument list
 in include(::ASCIIString) at ./boot.jl:233
 in include_from_node1(::ASCIIString) at ./loading.jl:426
 in eval(::Module, ::Any) at ./boot.jl:236
while loading /home/kristoffer/Documents/test.jl, in expression starting on line 4
```

Is there anyway the line info could be a bit more specific than just the line of the start of the enclosing function?

**Comment by yuyichao** [comment]
Dup of https://github.com/JuliaLang/julia/issues/15643

**Comment by KristofferC** [comment]
That one has the title "bad stack trace for errors in macros". This example has no macros. Should the title in #15643 be updates?

**Comment by JeffBezanson** [comment]
Looks like a different issue to me. This one is about syntax error locations.

**Comment by Keno** [comment]
Or you could just use JuliaParser ;):
<img width="270" alt="screen shot 2016-05-02 at 12 39 25 pm" src="https://cloud.githubusercontent.com/assets/1291671/14960499/efb51f5e-1062-11e6-910e-e0664b6dcb8f.png">

**Comment by KristofferC** [comment]
Wow, that is some clang beauty right there.

**Comment by Keno** [comment]
The README has instructions: https://github.com/JuliaLang/JuliaParser.jl#using-juliaparser-as-your-primary-parser

**Comment by KristofferC** [comment]
Here is another example from a type I made:

```jl
# test.jl
module Test

function filler()
    blabla
end

typealias a = Float64

end # module
```

```jl
julia> include("/home/kristoffer/Documents/test.jl")
ERROR: LoadError: syntax: unexpected "="
 in include_from_node1(::String) at ./loading.jl:426
 in eval(::Module, ::Any) at ./boot.jl:230
while loading /home/kristoffer/Documents/test.jl, in expression starting on line 1
```

**Comment by KristofferC** [comment]
Another one:

```
module M

function f()
	while true
		if x == 0 && y == 0
			a = 1
		if x == 1 && z == 1
			a = 2
		end
	end
end

end
```

This gives `ERROR: syntax: incomplete: "module" at REPL[44]:1 requires end`. For a large modules this is a quite unhelpful error and I usually end up pasting the whole module in the REPL and see where the latest completely parsed expression got completed. Maybe the Error message could show this as well?

**Comment by StefanKarpinski** [comment]
I think the best way to give better error messages is probably to consider indentation. To me that suggests bailing out on a parse when there's an error and reparsing the input with a parser that is slower and tracks extra information like indentation and gives really high quality messages.

**Comment by vtjnash** [comment]
Would be awesome if we could have https://github.com/KristofferC/OhMyREPL.jl replace the default `on_done` / `parse_input_line` function with one that uses that fancier parser (https://github.com/ZacLN/CSTParser.jl) hint hint hint :) :) :)

**Comment by Keno** [comment]
Already in progress: https://github.com/Keno/FancyDiagnostics.jl

**Comment by felixcremer** [comment]
This is closed in julia 1.5. The first example gives:

```julia
julia> include("test.jl")
ERROR: LoadError: syntax: missing comma or ) in argument list
Stacktrace:
 [1] top-level scope at /home/user/Documents/Programmieren/julia_projects/tmp/test.jl:11
 [2] include(::String) at ./client.jl:457
 [3] top-level scope at REPL[1]:1
in expression starting at /home/userDocuments/Programmieren/julia_projects/tmp/test.jl:11
```
This is the line directly after the missing comma.

**Comment by laborg** [comment]
JuliaSyntax handles the first problem:
```
julia> include("bug.jl")
ERROR: LoadError: ParseError:
# Error @ /home/gerhard/bug.jl:11:5
    f(a,(b,) # :(
    f(a,(b,))
#   └─────┘ ── Expected `)`
Stacktrace:
 [1] top-level scope
   @ ~/bug.jl:11
 [2] include(fname::String)
   @ Base.MainInclude ./client.jl:496
 [3] top-level scope
   @ REPL[10]:1
in expression starting at /home/gerhard/bug.jl:11
```
The missing end message is still not optimal. Not sure if there is actually a way to improve this.
